man/aam2mail.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH aam2mail 1 "March 2012" Linux "User Manuals"
.SH NAME
aam2mail \- Anonymous shared mailbox scanner
.SH SYNOPSIS
.B aam2mail
{ --start | --stop | --restart | --help }
.SH DESCRIPTION
.B aam2mail
extracts messages, by subject, from a Usenet Newsgroup. For many years the
Newsgroup alt.anonymous.messages has been used as a shared mailbox for users of
anonymous messaging solutions.  Messages for multiple users are delivered to
the mailbox and then extracted by their rightful recipient using Subject header
matching.
.P
Subjects can be identified in plain-text but more frequently a form of encoding
is employed. This encoding prevents messages being linked to a common recipient
by the Subject header's content.  There are two supported encoding methods,
known as eSub and hSub.  They both work by generating a hash consisting of a
secret key and public information contained within the message's Subject
header. When a collision occurs, the message is known to belong to the holder
of the secret.
.P
.B aam2mail
allows the user to define multiple sources (Newsservers) for retrieval of
messages from alt.anonymous.messages. Messages are then read from all the
sources, and scanned for Subject header collisions.  Any number of plain-text,
eSub or hSub secrets can be supplied for matching purposes.  Once a message is
identifed, it is stored locally in either \fIMaildir\fP or \fImbox\fP format
for decryption by a standard MUA, such as \fBMutt\fP.
.SH OPTIONS
.TP
.B --config
Override the default ~/.aam2mail configuration file.
.TP
.B --start
Start aam2mail as a daemon
.TP
.B --stop\ \ 
Stop the aam2mail daemon. No further actions will be performed until the
daemon is restarted or aam2mail is manually run.
.TP
.B --restart
Restart the aam2mail daemon.  The daemon should be restarted after a
configuration change or a logrotate.
.TP
.B --help\ 
Display a short help message
.SH FILES
This section describes the user-configurable files in \fBaam2mail\fP.
.TP
.B .aam2mailrc
This is the main configuration file.  It defines all the options relating to
the function of aam2mail, including the options to override the locations of
all the other configuration files.  By default \fB.aam2mailrc\fP resides in
$HOME.  The path and filename can be overridden with the \fB--config\fP option
or by setting the Environment Variable \fBAAM2MAIL\fP.
.P
.IP
The \fB.aam2mailrc\fP file is only required if a default option needs to be
overridden.  A sample config file \fBconfig.sample\fP is provided with the
distribution and contains details of each supported option.
Default=$HOME/.aam2mailrc
.TP
.B servers
The servers file contains a list of the full-qualified hostnames for the
newsservers that \fBaam2mail\fP should download from. One newsserver is
normally sufficient but more will reduce the minor risk of lost messages.
.P
.IP
.B aam2mail
will attempt to deduplicate downloading of messages when working with multiple
newsservers.  This is most effective if the newsservers are synchronized. I.e.
The same messages are being read from each server during the same operation.
This may not be the case if \fBfetch_limit\fP is less than the number of
messages on the server and \fBprocess_backlog\fP is true.
Default=$HOME/aam2mail/etc/servers
.P
.TP
.B subject_text
Contains a list of the plain-text subjects the user would like to retrieve from
the newsserver. This file is case-sensitive.
Default=$HOME/aam2mail/etc/subject_text
.P
.TP
.B subject_hsub
Contains a list of the hSub secrets that \fBaam2mail\fP should generate
collisions for when looking for subject matches.
Default=$HOME/aam2mail/etc/subject_hsub
.P
.TP
.B subject_esub
Contains a list of eSub subjects and keys for generating collisions with
subject headers. Unlike the other subject_files, esub requires two parameters
per line.  These should be seperated with a \fBTab\fP character.
Default=$HOME/aam2mail/etc/subject_esub
.SH ENVIRONMENT
.TP
.B AAM2MAIL
Overrides the default path ($HOME/.aam2mailrc) for the master configuration
file.  This setting can be overridden with the \fB--config\fP option.
.SH EXAMPLES
.TP
.B aam2mail --start
Start \fBaam2mail\fP in non-interactive, daemon mode.
.TP
.B aam2mail --config=/etc/aam2mail
Start \fBaam2mail\fP in the current console session and use \fI/etc/aam2mail\fP
as the configuration file.


CHANGES
v0.1, 13-Feb-2012 -- Initial release.

v0.2, xx-xxxx-2012 -- Fixes
o Correct a manpage entry that states subject_plain instead of subject_text.


aam2mail.py
#!/usr/bin/python
#
# vim: tabstop=4 expandtab shiftwidth=4 autoindent
#
# Copyright (C) 2012 Steve Crook <steve@mixmin.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTIBILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program.  If not, see <http://www.gnu.org/licenses/>.

import os.path
import mailbox
import sys
import nntplib
import socket
import logging
from Config import options
from Config import config
from time import sleep
from daemon import Daemon

class MyDaemon(Daemon):
    def run(self):
        while True:
            aam.main()
            interval = config.getint('usenet', 'fetch_interval')
            sleep(interval)

class aam():
    def __init__(self):
        # This section defines what type of Subjects we're interested in.  The
        # choices are plain text, hsub and esub.  These are marked for
        # processing if the corresponding config file exists and contains
        # entries.
        do_text = False
        do_hsub = False
        do_esub = False

        subj_list = self.file2list(os.path.join(config.get('paths', 'etc'),
                                                "subject_text"))
        if subj_list:
            do_text = True
            self.subj_list = subj_list
            logmsg = "Checking %s plain text Subjects" % len(subj_list)
            logging.info(logmsg)
        hsub_list = self.file2list(os.path.join(config.get('paths', 'etc'),
                                                "subject_hsub"))
        if hsub_list:
            do_hsub = True
            import hsub
            self.hsub = hsub.hsub()
            self.hsub_list = hsub_list
            logging.info("Checking %s hSub Subjects" % len(hsub_list))
        esub_list = self.file2list(os.path.join(config.get('paths', 'etc'),
                                                "subject_esub"))
        if esub_list:
            do_esub = True
            import esub
            self.esub = esub.esub()
            self.esub_list = esub_list
            logging.info("Checking %s eSub Subjects" % len(esub_list))

        if not do_text and not do_hsub and not do_esub:
            errmsg = "No text, hsub or esub Subjects defined. Aborting."
            logging.error(errmsg)
            sys.exit(1)

        # Populate the himarks dict and sync it with our etc/servers text file.
        server_file = os.path.join(config.get('paths', 'etc'), "servers")
        himark_file = os.path.join(config.get('paths', 'spool'), "servers")
        servers = self.file2list(server_file)
        himarks = self.file2dict(himark_file, numeric = True)
        # Add missing servers
        for server in servers:
            if server not in himarks:
                himarks[server] = 0
        # Delete unwanted servers
        # We have to take a copy of the keys as we can't delete keys that
        # we're looping over.
        hikeys = himarks.keys()
        for server in hikeys:
            if server not in servers:
                del himarks[server]
        # Set the scope on our variables
        self.do_text = do_text
        self.do_hsub = do_hsub
        self.do_esub = do_esub
        self.himarks = himarks
        self.himark_file = himark_file

    def file2list(self, filename):
        """Read a file and return each line as a list item. Note, if the file
        doesn't exist, this function will return an empty list."""
        items = []
        if os.path.isfile(filename):
            readlist = open(filename, 'r')
            for line in readlist:
                # This beast just strips comments from read lines.
                content = line.split('#', 1)[0].rstrip()
                # If a line starts with a #, content will be length zero
                if len(content) > 0:
                    items.append(content)
            readlist.close()
        return items

    def file2dict(self, filename, numeric = False):
        """Read a file and split each line at the first space encountered. The
        first element is the key, the rest is the content. If numeric is True
        then only integer values will be acccepted."""
        d = {}
        for line in self.file2list(filename):
            fields = line.split(" ", 1)
            k = fields[0]
            c = fields[1]
            if numeric:
                try:
                    c = int(c)
                except ValueError:
                    c = 0
                d[k] = c
            else:
                d[k] = c.strip()
        return d

    def dict2file(self, filename, d):
        "Write a dictionary to a text file"
        f = open(filename, 'w')
        for k in d:
            # In theory we shouldn't need to strip this, but in practice,
            # theory and practice are frequently different.
            f.write("%s %s\n" % (k.strip(), d[k]))
        f.close()

    def list2multi_line_string(self, list):
        """Take a list and return it as a multi-line string."""
        string = ""
        for line in list:
            string += line + "\n"
        return string

    def mail_headers(self, msgid, sender, date):
        """Add some basic headers to our message body so it can be processed in
        Maildir format."""
        payload = "Date: %s\n" % date
        payload += "From: %s\n" % sender
        payload += "Subject: Nym message from %s\n" % sender
        payload += "Message-ID: %s\n" % msgid
        return payload

    def get_range(self, sfirst, slast, himark):
        """Return a range of article numbers we should process.  We determine
        this by comparing the articles available on the server with those (if
        known) from our previous use of this server."""
        first = int(sfirst)
        last = int(slast)
        limit = config.getint('usenet', 'fetch_limit')

        if himark >= first and himark <= last:
            # This is the normal state of affairs. Our himark lies between the
            # server's high and low. We update first with our last himark plus
            # one (We don't want our last himark message, hence the +1).
            first = himark + 1

        # Check we aren't receiving more messages than the configured limit.
        howmany = last - first
        if howmany > limit:
            logmes = "%s: There are %s unread messages. " % (self.server,
                                                             howmany)
            logmes += "Only processing %s, due to " % limit
            logmes += "configured fetch_limit."
            logging.warn(logmes)
            last = first + (limit - 1)
            #first = last - (limit - 1)
        else:
            logging.info("Fetching all %s available articles." % (howmany + 1))
        return str(first), str(last)

    def xover(self, spool_file):
        # group returns: response, count, first, last, name
        try:
            resp, grpcount, grpfirst, \
            grplast, grpname = self.news.group(config.get('usenet',
                                                          'newsgroup'))
        except nntplib.NNTPTemporaryError, e:
            logging.warn("%s: %s" % (self.server, e))
            return 0
        first, last = self.get_range(grpfirst,
                                     grplast,
                                     self.himarks[self.server])
        msgcnt = (int(last) - int(first)) + 1
        logmes = "%s: Processing %d messages" % (self.server, msgcnt)
        logging.debug(logmes)
        if msgcnt <= 0:
            return int(last)
        # The following xover line is often remarked out during testing as
        # this preserves a constant tmpfile.
        try:
            resp, foo = self.news.xover(first, last, spool_file)
        except nntplib.NNTPTemporaryError, e:
            logging.warn("%s: %s" % (self.server, e))
            return 0
        logging.debug("Xover responded with: %s" % resp)
        return int(last)

    def retrieve(self, spool_file):
        isopen_maildir = False
        isopen_mbox = False
        received = 0
        s = open(spool_file, "r")
        for ov in s:
            ov2 = ov.rstrip()
            items = ov2.split("\t")
            subject = items[1]
            sender = items[2]
            date = items[3]
            msgid = items[4]

            # Skip checking messages if we already have this Message-ID
            # Not a great check as it's only relevent if all servers we
            # pull from are in sync.
            if msgid in self.dedupe:
                logging.debug("%s: Already processed." % msgid)
                continue

            # Retreive the actual payload.  This is amazingly inefficient as we
            # could just get the ones we want but that's bad for anonymity.
            if config.getboolean('usenet', 'fetch_all'):
                body = self.getbody(msgid)
                if body is None: continue
            wanted = False
            if self.do_text and subject in self.subj_list:
                wanted = True
            if not wanted and self.do_hsub:
                for subj in self.hsub_list:
                    if self.hsub.check(subj, subject):
                        wanted = True
                        # break out of hsub matching. It can match only one.
                        break
            if not wanted and self.do_esub:
                for subj in self.esub_list:
                    # In our esub file, we assume Subject and Key are separated
                    # by a tab.
                    encsub, key = subj.split("\t", 1)
                    if self.esub.check(encsub, key, subject):
                        wanted = True
                        # break out of esub matching. It can match only one.
                        break
            if wanted:
                if not config.getboolean('usenet', 'fetch_all'):
                    body = self.getbody(msgid)
                    if body is None: continue
                headers = self.mail_headers(msgid, sender, date)
                msg = "%s\n%s" % (headers, body)
                # Create the message in the Maildir
                if config.getboolean('mailboxes', 'do_maildir'):
                    if not isopen_maildir:
                        maildir = mailbox.Maildir(config.get('mailboxes',
                                                             'maildir'),
                                                  create = True)
                        isopen_maildir = True
                    maildir.add(msg)
                 # If required, configure mbox processing
                if config.getboolean('mailboxes', 'do_mbox'):
                    if not isopen_mbox:
                        mbox = mailbox.mbox(config.get('mailboxes', 'mbox'),
                                            create = True)
                        isopen_mbox = True
                    mbox.add(msg)
                # Increment the received message count
                received += 1
            # Add the Message-ID to the dedupe list.  We won't need to examine
            # it again, whether we wanted it or not.
            self.dedupe.append(msgid)
        # If we opened maildir or mbox objects, we close them now.
        if isopen_maildir:
            maildir.close()
        if isopen_mbox:
            mbox.close()
        return received

    def getbody(self, msgid):
        # The tuple returned by nntp.body includes the actual body as a
        # list object as the fourth element, hence [3].
        try:
            body = self.list2multi_line_string(self.news.body(msgid)[3])
        except socket.error:
            logging.warn("Timeout retrieving %s" % msgid)
            return None
        except nntplib.NNTPTemporaryError:
            message = '%s: news.body returned a temporary ' % self.server
            message += 'error: %s.' % sys.exc_info()[1]
            logging.warn(message)
            return None
        except nntplib.NNTPPermanentError:
            message = '%s: news.body returned a permanent ' % self.server
            message += 'error: %s.' % sys.exc_info()[1]
            logging.warn(message)
            return None
        return body

    def main(self):
    # Dedupe maintains a list of Message-IDs so that duplicate messages aren't
    # created when we have multiple news servers defined.
        self.dedupe = []
        received = 0
        srv_count = 0
        # Main loop of servers to process starts here.
        for server in self.himarks:
            # We use the server name so broadly, it's worth scoping it out.
            self.server = server
            # Assign a temporary file name for storing xover data.  The name is
            # based on <tempdir>/<server>.tmp
            spool = os.path.join(config.get('paths', 'spool'),
                                 server + '.tmp')

            # Establish a connection with the newsserver.
            logging.debug("%s: Establishing connection." % server)
            try:
                self.news = nntplib.NNTP(server, readermode = True)
            except nntplib.NNTPTemporaryError, e:
                logging.warn('%s: Connection error: %s' % (server, e))
                continue
            except socket.gaierror, e:
                logging.warn('%s: Connection error: %s' % (server, e))
                continue
            himark = self.xover(spool)
            # Zero isn't a valid himark so our xover routine can return it
            # to indicate something went wrong.  Hopefully the something is
            # logged.
            if himark == 0:
                continue
            # If we get here, the assumption is that the server accepted our
            # connection and provided any messages we asked for.
            srv_count += 1
            #If the himark we get back from xover is the same as we supplied
            #it then there are no new messages to process.
            if himark == self.himarks[server]:
                # we didn't fetch any messages so move on to the next server
                logging.debug("%s: No new messages." % server)
                self.news.quit()
                continue
            # Now we retrieve messages from the servers and test them.
            received += self.retrieve(spool)
            self.himarks[server] = himark
            self.news.quit()
        if received > 0:
            logging.info("Received %d messages." % received)
        else:
            logging.debug("Received %d messages." % received)
        msg = "We processed %d unique messages " % len(self.dedupe)
        msg += "from %d servers " % srv_count
        msg += "(%d attempted)." % len(self.himarks)
        logging.debug(msg)
        # Write the revised server db
        logging.debug("Writing himarks to file.")
        self.dict2file(self.himark_file, self.himarks)

def init_logging():
    loglevels = {'debug': logging.DEBUG, 'info': logging.INFO,
                 'warn': logging.WARN, 'error': logging.ERROR}
    # No dynamic logfile name as we're running as a daemon
    logfile = os.path.join(config.get('paths', 'log'), 'aam2mail.log')
    logging.basicConfig(
        filename=logfile,
        level = loglevels[config.get('general', 'loglevel')],
        format = '%(asctime)s %(process)d %(levelname)s %(message)s',
        datefmt = '%Y-%m-%d %H:%M:%S')

if __name__ == "__main__":
    init_logging()
    aam = aam()
    pidfile = os.path.join(config.get('paths', 'run'), 'aam2mail.pid')
    errfile = os.path.join(config.get('paths', 'log'), 'aam2mail.err')
    daemon = MyDaemon(pidfile, '/dev/null', '/dev/null', errfile)
    # Process the start/stop args
    if options.start:
        logging.info('Started in Daemon mode')
        sys.stdout.write('aam2mail: Starting in Daemon Mode\n')
        daemon.start()
    elif options.stop:
        daemon.stop()
        logging.info("Stopping Daemon")
        sys.stdout.write('aam2mail stopped\n')
    elif options.restart:
        logging.info("Restarting Daemon")
        daemon.restart()
    else:
        # Run in console for testing
        daemon.run()
